File: Problem 70.cpp
class Node {
    public:
        vector<Node *>link;
        bool isEnd;

        Node() {
        link = vector<Node*>(26, nullptr);
        isEnd = false;
        }

        bool isHaving(char ch)
        {
            return link[ch-'a']!=NULL;
        }

        Node* nextLink(char ch)
        {
            return link[ch - 'a'];
        }
};

class Trie {
    Node* root;
public:
    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        int n = word.size();
        Node* node = root;
        for (int i = 0; i < n; i++) {
            if (! node->isHaving(word[i])) {
                node->link[word[i] - 'a'] = new Node();
            }
            node = node->nextLink(word[i]);
        }
        node->isEnd = true;
    }
    
    bool search(string word) {
        int n = word.size();
        Node* node = root;
        for (int i = 0; i < n; i++) {
            if (! node->isHaving(word[i])) return false;
           node = node->nextLink(word[i]);
        }
        return node->isEnd;
    }
    
    bool startsWith(string prefix) {
        int n = prefix.size();
        Node* node = root;
        for (int i = 0; i < n; i++) {
            if (! node->isHaving(prefix[i])) return false;
           node = node->nextLink(prefix[i]);
        }
        return true;
    }
};

File: Problem 70.py
class TrieNode:
    def __init__(self):
        self.link={}
        self.isEnd=False

class Trie:

    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        n = len(word)
        node = self.root
        for ch in word:
            if ch not in node.link:
                node.link[ch]= TrieNode()
            node = node.link.get(ch)
        node.isEnd= True

    def search(self, word: str) -> bool:
        n = len(word)
        node = self.root
        for ch in word:
            if ch not in node.link: return False
            node = node.link.get(ch)
        return node.isEnd
        

    def startsWith(self, prefix: str) -> bool:
        n = len(prefix)
        node = self.root
        for ch in prefix:
            if ch not in node.link: return False
            node = node.link.get(ch)
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)